File: 25/core/decryptor.py
# -*- coding: utf-8 -*-
"""
.sxc 文件解密模块
和 JS 端 AES-256-GCM 算法完全兼容，同一个密钥
解密全在内存，绝不落盘
"""
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _get_master_key() -> bytes:
    """和你 JS 端完全一样的密钥，打乱顺序防反编译"""
    part3 = b'wxaczxah'
    part1 = b'sxcccfgh'
    part4 = b's1172581'
    part2 = b'sscgxdsg'
    key_str = part1 + part2 + part3 + part4
    return key_str


def decrypt_sxc(encrypted_data: bytes) -> dict:
    """
    解密 .sxc 文件，内存中执行
    :param encrypted_data: .sxc 文件的二进制内容
    :return: 解密后的凭证 JSON（包含 cookies、envFingerprint）
    """
    header = b'SXC_ENC:'
    if not encrypted_data.startswith(header):
        raise ValueError("无效的 .sxc 加密文件")

    encrypted_b64 = encrypted_data[len(header):].decode('utf-8')
    encrypted = base64.b64decode(encrypted_b64)

    iv = encrypted[:12]
    ciphertext = encrypted[12:]

    key = _get_master_key()
    aesgcm = AESGCM(key)

    try:
        plaintext = aesgcm.decrypt(iv, ciphertext, associated_data=None)
    finally:
        del key, iv, ciphertext, encrypted_b64, encrypted

    import json
    try:
        credentials = json.loads(plaintext.decode('utf-8'))
    finally:
        del plaintext

    return credentials


def wipe_memory(obj):
    """尝试清零对象内存引用（降低残留时间，不能保证完全）"""
    try:
        if isinstance(obj, dict):
            for v in obj.values():
                wipe_memory(v)
            obj.clear()
        elif isinstance(obj, list):
            for v in obj:
                wipe_memory(v)
            obj.clear()
        elif isinstance(obj, bytes):
            for i in range(len(obj)):
                obj[i] = 0
    except Exception:
        pass

File: 32内置统计积分初始版/core/decryptor.js
/**
 * .sxc 文件解密模块
 * Node.js (Electron 主进程) 专用，使用 Node 内置 crypto 模拟 Web Crypto API
 * 算法：AES-256-GCM，和 Chrome 扩展端完全一致
 * 全在内存执行，绝不落盘
 */

// Node.js 原生 crypto（Electron 主进程环境保证可用）
const { webcrypto } = require('crypto');
const crypto = webcrypto;

// 模拟浏览器端的 atob/btoa
function atob(str) {
    return Buffer.from(str, 'base64').toString('binary');
}
function btoa(str) {
    return Buffer.from(str, 'binary').toString('base64');
}

/**
 * 获取 AES-256 解密密钥
 * 和扩展端 JS 完全一样的密钥，打乱顺序防反编译
 */
async function _getMasterKey() {
    const part3 = btoa('wxaczxah');
    const part1 = btoa('sxcccfgh');
    const part4 = btoa('s1172581');
    const part2 = btoa('sscgxdsg');
    const keyStr = atob(part1) + atob(part2) + atob(part3) + atob(part4);
    return new TextEncoder().encode(keyStr);
}

/**
 * 解密 .sxc 加密文件
 * @param {Buffer|Uint8Array} encryptedData - .sxc 文件二进制内容
 * @returns {Promise<Object>} 解密后的凭证（cookies, envFingerprint 等）
 */
async function decryptSxc(encryptedData) {
    // 统一转为 Uint8Array
    const raw = Buffer.isBuffer(encryptedData)
        ? new Uint8Array(encryptedData)
        : encryptedData;

    // 1. 解析头部
    const headerBytes = new Uint8Array([0x53, 0x58, 0x43, 0x5f, 0x45, 0x4e, 0x43, 0x3a]); // "SXC_ENC:"
    const headerMatches = raw.length > 8 &&
        headerBytes.every((b, i) => raw[i] === b);

    if (!headerMatches) {
        throw new Error("无效的 .sxc 加密文件");
    }

    // 2. 解码 Base64 部分
    const b64Str = Buffer.from(raw.slice(8)).toString('utf8');
    const encrypted = Uint8Array.from(Buffer.from(b64Str, 'base64'));

    // 3. 拆分 IV（前 12 字节）和密文
    const iv = encrypted.slice(0, 12);
    const ciphertext = encrypted.slice(12);

    // 4. AES-256-GCM 解密
    const key = await _getMasterKey();
    const aesGcm = await crypto.subtle.importKey(
        'raw', key, { name: 'AES-GCM' }, false, ['decrypt']
    );

    let plaintext;
    try {
        plaintext = await crypto.subtle.decrypt(
            { name: 'AES-GCM', iv: iv },
            aesGcm,
            ciphertext
        );
    } finally {
        // 内存清零
        key.fill(0);
        iv.fill(0);
        ciphertext.fill(0);
    }

    // 5. 解析 JSON
    const credentials = JSON.parse(Buffer.from(plaintext).toString('utf8'));

    return credentials;
}

module.exports = { decryptSxc };
